HW2/P4/driver.py

```diff
@@ -15,18 +15,44 @@
 from timer import Timer
 import threading
 
-def py_median_3x3(image, iterations=10, num_threads=1):
+def py_median_3x3(image, iterations=10, num_threads=4):
     ''' repeatedly filter with a 3x3 median '''
     tmpA = image.copy()
     tmpB = np.empty_like(tmpA)
 
-    for i in range(iterations):
-        filtering.median_3x3(tmpA, tmpB, 0, 1)
-        # swap direction of filtering
-        tmpA, tmpB = tmpB, tmpA
+    # Create an event that represents each call to filter_image (one event for each iteration in each thread)
+    events = [[threading.Event() for i in xrange(iterations)] for j in xrange(num_threads)]
+
+    # Generate list of threads and start them
+    threads = []
+    for i in xrange(num_threads):
+        threads.append(threading.Thread(target=filter_image, args=(events, i, num_threads, tmpA, tmpB, iterations)))
+        threads[i].start()
+
+    # Join threads
+    for thread in threads:
+        thread.join()
 
     return tmpA
 
+# Function called by each thread to perform filtering
+def filter_image(events, thread_idx, num_threads, tmpA, tmpB, iterations):
+    for i in xrange(iterations):
+        # Ignore for i = 0 because we only wait on neighboring threads to complete i-1 iterations for i > 0
+        # Ignore for num_threads <= 1 because no neighboring threads exist
+        if num_threads > 1 and i > 0:
+            # Need to wait for neighboring threads to complete i-1 iterations
+            events[(thread_idx-1)%num_threads][i-1].wait()
+            events[(thread_idx+1)%num_threads][i-1].wait()
```

Comment:
I'm not sure but is the mod redundant here since thread_idx will be <=num_threads-1.